crates/vm/src/memory.rs

```diff
@@ -0,0 +1,117 @@
+use essential_types::Word;
+
+use crate::error::MemoryError;
+
+#[cfg(test)]
+mod tests;
+
+#[derive(Clone, Default, Debug, PartialEq)]
+/// Memory for temporary storage of words.
+pub struct Memory(Vec<Word>);
+
+impl Memory {
+    /// The maximum number of words that can be stored in memory.
+    pub const SIZE_LIMIT: usize = 1024 * 10;
+
+    /// Create a new temporary memory instance.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Allocate more memory to the end of this memory.
+    pub fn alloc(&mut self, size: Word) -> Result<(), MemoryError> {
+        let size = usize::try_from(size).map_err(|_| MemoryError::Overflow)?;
+        let new_size = self
+            .0
+            .len()
+            .checked_add(size)
+            .ok_or(MemoryError::Overflow)?;
+        if new_size > Self::SIZE_LIMIT {
+            return Err(MemoryError::Overflow);
+        }
+        self.0.resize(new_size, 0);
+        Ok(())
+    }
+
+    /// Store a word at the given address.
+    pub fn store(&mut self, address: Word, value: Word) -> Result<(), MemoryError> {
+        let index = usize::try_from(address).map_err(|_| MemoryError::IndexOutOfBounds)?;
+        *self.0.get_mut(index).ok_or(MemoryError::IndexOutOfBounds)? = value;
+        Ok(())
+    }
+
+    /// Load a word from the given address.
+    pub fn load(&mut self, address: Word) -> Result<Word, MemoryError> {
+        let index = usize::try_from(address).map_err(|_| MemoryError::IndexOutOfBounds)?;
+        Ok(*self.0.get(index).ok_or(MemoryError::IndexOutOfBounds)?)
+    }
+
+    /// Store a range of words starting at the given address.
+    pub fn store_range(&mut self, address: Word, values: &[Word]) -> Result<(), MemoryError> {
+        let address = usize::try_from(address).map_err(|_| MemoryError::IndexOutOfBounds)?;
+        let end = address
+            .checked_add(values.len())
+            .ok_or(MemoryError::Overflow)?;
+        if end > self.0.len() {
+            return Err(MemoryError::IndexOutOfBounds);
+        }
+        self.0[address..end].copy_from_slice(values);
+        Ok(())
+    }
+
+    /// Load a range of words starting at the given address.
+    pub fn load_range(&mut self, address: Word, size: Word) -> Result<Vec<Word>, MemoryError> {
+        let address = usize::try_from(address).map_err(|_| MemoryError::IndexOutOfBounds)?;
+        let size = usize::try_from(size).map_err(|_| MemoryError::Overflow)?;
+        let end = address.checked_add(size).ok_or(MemoryError::Overflow)?;
+        if end > self.0.len() {
+            return Err(MemoryError::IndexOutOfBounds);
+        }
+        Ok(self.0[address..end].to_vec())
+    }
+
+    /// Free some memory from an index to the end of this memory.
+    pub fn free(&mut self, address: Word) -> Result<(), MemoryError> {
+        let index = usize::try_from(address).map_err(|_| MemoryError::IndexOutOfBounds)?;
+        if index >= self.0.len() {
```

Comment:
I think we should allow `index` to be equal to `self.0.len()` which would make this a no-op. I'm in a situation where I have some data of size `N` in memory and potentially other stuff (I don't know if they're there or not!) and I want to clear everything except for the first `N` words. I want to be able to do:
```
PUSH N
FREE
```
Just changing this change to `if index > self.0.len()` makes this work. What do you think?

Reply:
Sgtm! I'll push a commit for this and check the spec docs

Reply:
Opened PR https://github.com/essential-contributions/essential-base/pull/239